airflow/models/dag.py

```diff
@@ -1190,13 +1189,8 @@ def add_task(self, task):
             task.end_date = min(task.end_date, self.end_date)
 
         if task.task_id in self.task_dict:
```

Comment:
```
dag =  DAG("test_dag", start_date=DEFAULT_DATE):
dag.add_task(t1)
dag.add_task(t1)
```
Will this code still work?  In my opinion, re-adding the same task should not cause an exception.

Reply:
Adding the same _object_ is slightly different to adding a new task with a duplicate id, but I can't think of any reason good you'd want to do the case as Kamil showed in practice, so I'd go for just making it an error

Reply:
Currently all DAGs are broken.
For example, in example_python_operator we initialize new task using following code
```
run_this = PythonOperator(
    task_id='print_the_context',
    python_callable=print_context,
    dag=dag,
)
```
This code refers to the DAG property being set. 
https://github.com/apache/airflow/blob/master/airflow/models/baseoperator.py#L410
This property call BaseOperator.add_task method
https://github.com/apache/airflow/blob/master/airflow/models/baseoperator.py#L532
This method set the dag property
https://github.com/apache/airflow/blob/master/airflow/models/dag.py#L1202
This property call BaseOperator.add_task method
https://github.com/apache/airflow/blob/master/airflow/models/baseoperator.py#L532
This flow causes that we return to the same method again, so the first condition is met, so we have exception. 

so we need to allow the option of adding the same task multiple times.

Reply:
My solution:
```diff
From de7c6a66108cccdcd23927ad0e821f5150741602 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kamil=20Bregu=C5=82a?= <kamil.bregula@polidea.com>
Date: Tue, 12 Nov 2019 17:07:04 +0100
Subject: fixup! [AIRFLOW-5880] Enforce unique task ids

---
 airflow/models/dag.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/airflow/models/dag.py b/airflow/models/dag.py
index 57091ac7d..13c121dcc 100644
--- a/airflow/models/dag.py
+++ b/airflow/models/dag.py
@@ -1170,6 +1170,10 @@ class DAG(BaseDag, LoggingMixin):
         :param task: the task you want to add
         :type task: task
         """
+        # if the task has already been added. do anything.
+        if task.task_id in self.task_dict and self.task_dict[task.task_id] == task:
+            return
+
         if not self.start_date and not task.start_date:
             raise AirflowException("Task is missing the start_date parameter")
         # if the task has no start date, assign it the same as the DAG
-- 
2.20.1
```

Reply:
Adding the same task should be fine. Just added another commit with the above solution